cpp/src/neighbors/detail/quantization.cuh

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuvs/neighbors/quantization.hpp>
+#include <raft/core/operators.hpp>
+#include <raft/linalg/unary_op.cuh>
+#include <raft/random/rng.cuh>
+#include <raft/random/sample_without_replacement.cuh>
+#include <thrust/execution_policy.h>
+#include <thrust/sort.h>
+#include <thrust/system/omp/execution_policy.h>
+
+namespace cuvs::neighbors::detail {
+
+template <typename T, typename QuantI = int8_t, typename TempT = double, typename TempI = int64_t>
+raft::device_matrix<QuantI, int64_t> scalar_quantize(
+  raft::resources const& res,
+  cuvs::neighbors::quantization::params& params,
+  raft::device_matrix_view<const T, int64_t> dataset)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(res);
+
+  constexpr TempI q_type_min   = static_cast<TempI>(std::numeric_limits<QuantI>::min());
+  constexpr TempI q_type_max   = static_cast<TempI>(std::numeric_limits<QuantI>::max());
+  constexpr TempI range_q_type = q_type_max - q_type_min + TempI(1);
+
+  size_t n_elements = dataset.extent(0) * dataset.extent(1);
+
+  // conditional: search for quantiles / min / max
+  if (!params.is_computed) {
+    ASSERT(params.quantile > 0.5 && params.quantile <= 1.0,
+           "quantile for scalar quantization needs to be within (.5, 1]");
+
+    double quantile_inv = 1.0 / params.quantile;
+
+    // select subsample
+    int seed                     = 137;
+    constexpr size_t num_samples = 10000;
```

Comment:
For a dataset that has dim=1k , this would correspond to 10 vectors. I suggest to increase this significantly.